### PR TITLE
Add support for typechecking IRs in full relational IR

### PR DESCRIPTION
**Pull request by tpoterba**
I don't add typecheck rules for the relational nodes themselves,
but this is now very easy to add.

### CODE REVIEW

**Comment by tpoterba** [comment]
If this doesn't catch at least one bug I'll be surprised.

File: hail/src/main/scala/is/hail/expr/ir/TypeCheck.scala [hunk 1]
```diff
@@ -9,27 +8,113 @@ object TypeCheck {
...
     } catch {
       case e: Throwable => fatal(s"Error while typechecking IR:\n${ Pretty(ir) }", e)
     }
   }
 
-  private def _apply(ir: IR, env: Env[Type], aggEnv: Option[Env[Type]]) {
+  private val empty = Env.empty[Type]
+
+  private def checkBlockMatrix(ir: BlockMatrixIR): Unit = {
+    // TODO: fill in when BM is more stable
```
Comment (jigold) [inline hail/src/main/scala/is/hail/expr/ir/TypeCheck.scala]: @daniel-goldstein Just making sure you are aware of this.

**Comment by jigold** [comment]
I don't add typecheck rules for the relational nodes themselves,

> I don't add typecheck rules for the relational nodes themselves,

I'm not sure what you mean by this. Also, were we just not type checking MatrixIR, etc. before? Was there a reason for this? I thought @patrick-schultz was working on something to do with type checking.

**Comment by daniel-goldstein** [comment]
> I'm not sure what you mean by this. Also, were we just not type checking MatrixIR, etc. before? Was there a reason for this? I thought @patrick-schultz was working on something to do with type checking.

Currently, typechecking assertions happen in the constructors of relational IRs. This will consolidate typechecking nicely. The expression problem strikes again!

**Comment by tpoterba** [comment]
I have a better approach.

**Pull request closed without merging**